### ISSUE TITLE
FEATURE REQUEST: Add the ability to apply multiple layouts

### ISSUE THREAD

**Issue by dking02** [Ross-Patterson/Portals-Desktop-Organization#139]
When a saved layout that includes a portal that already exists in the current layout is being applied (Setup > Home Tab > Layouts > Apply Layout), Portals will pop-up an error message “Portal already exists. Can’t create portal”. Add the name of the affected portal to the message so that the user can properly assess the situation and take corrective action if applicable. 

Note: Portals will not overwrite (replace) a portal that already exists in the current layout. The portal in the current layout must first be closed. If a portal already exists, the affected portal in the layout being applied will be skipped. Also note that all portals on the desktop do not need to be applied in a single layout; multiple layouts can be applied sequentially offering additional opportunities for this situation to occur.

**Comment by Ross-Patterson** [comment]
Can you give more info on the exact setup & steps?

Here is what actually happens when a layout is applied:
1. All current Portals are closed.
2. The portals in the saved layout are created.

So not able to replicate the scenario where you get the "Portal already exists" scenario.

**Comment by dking02** [comment]
Steps to reproduce:
1. Close all portals to clear the current layout.
2. Load a saved layout which includes more than one portal.
3. Close all portals just created in the current layout --- except for one.
4. Now load the same layout as in step 1.

You will get the error message because the one portal you did not close in step 3 still exists.

The current implementation allows multiple layouts to be applied in succession and that is not possible if you are automatically closing all portals before applying another layout as you indicate.  Allowing multiple layouts enables the final current layout to be created from several smaller or reusable layouts without necessarily saving the final current layout or losing all your work,

In the current wording of #139, I recommended just including the name of the portal in question so the user would know which one was causing the problem and could take action to resolve the duplication. I refrained from suggesting in #139 that you ask the user whether to skip the portal (like it currently occurs) or replace the portal in the current layout with the definition in the layout being applied - sort of like Windows does when you attempt to overwrite a file. If the specification of the portal in the 
portal being applied was the one actually desired, the user would probably have to start over if the existing portal is not replaced.

**Comment by dking02** [comment]
This issue has been categorized as “Fixed” in v3.0.2.1 but the requested feature has not been implemented. Instead, it appears that a “Close All Portals” is now being executed automatically before applying each layout. This eliminates the ability to apply several layouts in sequence to build up the current layout and therefore the need for the requested feature.  

A relatively simple alternative would be to leave the existing capability to apply multiple layouts by making it an option (or experimental option) in App Settings with the default setting being to close all portals before applying a layout as changed in v3.0.2.1.  The removal of this feature means that all portals on the desktop must be applied in a single saved layout, one for each possible arrangement or augmented with manual creation and closing/hiding of portals. Allowing multiple layouts supports scenarios, such as applying separate layouts independently for displays 1, 2 and 3 or for Projects A, B, C, D without having to close out every portal on the desktop or having to create separate saved layouts for the entire desktop for each possible combination. Building up complex arrangements of portals from smaller layouts also reduces the amount of work to revise layouts and the likelihood of accidental loss of large amounts of changes when revising layouts.

**Comment by Ross-Patterson** [comment]
Nothing was changed though, this is why I was struggling to replicate the scenario you describe above. "Close All Portals" has always been called before applying a layout. I had never even thought about compounding layouts.

Saying that, I think its an interesting idea, and it would be very easy to implement a setting to skip the close all portals step when applying a layout. For any conflicts with existing portals, I could just skip them and show 1 message at the end of the process showing all portals that were skipped due to conflicts.

**Comment by dking02** [comment]
Nothing has changed? A "Fixed" label was applied to this issue when v3.0.2.1 was released to beta on 24 May. That is why I checked it out against that version.

It turns out that compounded layouts stopped working in v3.0.2.0 (beta). I only had that version running for a few hours back on 14 May to check the 3 fixes and 2 enhancements in that version. I did not check whether compounded layouts still worked in that version and reverted back to v3.0.1.0 (production) until v3.0.2.1 was released. I reinstalled v3.0.2.0 today (26 May) and discovered that compound layouts had initially stopped working in that release.

This issue (#139) was originally written against v3.0.1.0. I reinstalled v3.0.1.0 today and have confirmed that compound layouts does work in that version. As I recall, it also worked in v2.1.0.6 (Oct 2022). What I am now observing in v3.0.1.0 on closer look is that when a new layout is applied, Portals issues a warning message that the current layout will be lost if not saved, but it actually does not close the portals. This allows portals in the new layout to be added to the current layout. This failure to close the existing portals was apparently a latent error. Given the current lack of documentation, I had assumed compounded layouts was intended to be supported because it is actually useful. If you save individual portals or small groups of portals in layouts, they can be added fully customized to the desktop as needed (rather than having to load a larger layout for the full desktop). They can also be used and reused to built up larger layouts which also can be saved..

If you recall, there was regression in v3.0.1.0 where the "Close All Portals" button did not work which I reported in issue #136 after noticing the problem when upgrading from v2.1.0.6. That issue was fixed in v3.0.2.0.  I suspect that changes made in fixing that issue may have also fixed the latent problem with not closing the current portals when applying a new layout.

**Comment by Ross-Patterson** [comment]
You are right, it just clicked to me. The close all portals API was not working, so when applying a layout previously it was supposed to be closing all portals but it wasn't due to this. After fixing #136, the layouts also started working properly as per my original intention.

I marked it as fixed due to implementing the name of the affected portal into the error message, but this message won't actually appear anymore due to fixing the actual problem.

It is now working as originally intended, in 3.0.2.1. The behaviour you were seeing was incorrect but actually turned out useful for you.

We can rename this feature request to be related to add the ability to compound layouts.

**Comment by dking02** [comment]
After some additional consideration, I think it would be much better to add a separate “Apply Additional Layout” column to the Layouts table rather than to add a new global setting in App Behavior to specify whether a “Close All Portals” command should be issued when applying layouts.

Recommendation: The existing “Apply Layout” column would remain unchanged. It would include issuing a “Close All Portals” before applying the layout. The new second “Apply” column would be labeled “Apply Additional Layout” and it would not issue the “Close All Portals” before applying the layout in order to allow multiple layouts. 

If someone is only applying one layout, they click the “Apply Layout” column. If someone wants to apply multiple layouts, they click the “Apply Layout” column for the first layout and the “Apply Additional Layout” for any subsequent layouts.

I think this is a much clearer user interface as it is very explicit about what is being done and there is no need for the user to access other pages in Settings to manually issue a Close All Portals or change a global setting to apply multiple layouts. It also has the advantage of making application of layouts independent of the state of a global setting. This will make it easier in the future to script the application of layouts since there will be no need to have to check the state of the global setting or possibly have to toggle its state in the script.

**Comment by Ross-Patterson** [comment]
This is not in my roadmap and I have no plans to implement applying multiple layouts
